### PR TITLE
#167377536: Signup validation middleware

### PR DESCRIPTION
**Pull request by Jonathan4github**
#### What does this PR do?
This pull request implements a signup validation middleware
#### Description of Task to be completed

- create a validator folder
- implement a signup validation functionality
- write tests for signup validation

#### How should this be manually tested?

- git clone branch
- cd to project directory
- run `npm install` && run `npm test`

#### What are the relevant pivotal tracker stories?

[#167377536](https://pivotaltracker.com/story/show/167377536)

#### Screenshots (if appropriate)
N/A

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
```
Comment (Merkll) [inline .vscode/launch.json]: add this file to .gitignore

File: src/middlewares/validators/AuthValidator.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import { isLength, isEmail } from 'validator';
+
+export default {
+  /**
+   * @param {object} req - request object
+   * @param {object} res - request object
```
Comment (Bastieno) [inline src/middlewares/validators/AuthValidator.js]: Fixed this:
res - response object

Comment (Jonathan4github) [reply src/middlewares/validators/AuthValidator.js]: > Fixed this:
> res - response object

Noted

**Review by Bastieno — APPROVED** [review]
This looks good!

File: src/middlewares/validators/AuthValidator.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      errorMessage.username = `Username ${nameErrorMessage}`;
+    }
+    if (!isEmail(email)) {
+      errorMessage.email = 'Enter a valid email address';
+    }
+    if (!isLength(password, { min: 8 })) {
+      errorMessage.password = 'Password must be greater than 8 characters';
+    }
+    if (!/[aA-zZ]+[0-9]+/g.test(password) && /[^a-zA-Z0-9]*/g.test(password)) {
+      errorMessage.characters = 'Password must be alphanumeric characters';
```
Comment (Merkll) [inline src/middlewares/validators/AuthValidator.js]: can we look into express-validator or joi?

File: tests/auth.test.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        res.should.have.status(400);
+        res.body.error.should.eql('All or some of the field is/are undefined');
+      });
+    done();
+  });
+  it('should return error message when first name is not valid', (done) => {
+    chai
+      .request(app)
+      .post(signupRoute)
+      .send(invalidFirstname)
```
Comment (Merkll) [inline tests/auth.test.js]: can use asyn/await

**Review by Merkll — APPROVED** [review]

File: package.json [hunk 5]
```diff
@@ -18,8 +18,9 @@
     "sequelize-cli:es6": "babel-node ./node_modules/.bin/sequelize $*",
     "start": "node dist/index.js",
     "start:dev": "cross-env DEBUG=dev nodemon --exec babel-node src/index.js",
+    "start:debug": "cross-env DEBUG=dev nodemon --inspect --exec babel-node src/index.js",
     "sync-env": "sync-dotenv",
-    "test": "mocha --require @babel/register --require regenerator-runtime/runtime tests/test.js --exit",
+    "test": "mocha --require @babel/register --require regenerator-runtime tests/**test.js --exit",
```
Comment (iverenshaguy) [inline package.json]: Please fix this line

File: src/middlewares/AuthValidator.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
+
+/**
+  * @param {object} request - request object
```
Comment (iverenshaguy) [inline src/middlewares/AuthValidator.js]: Ensure your JSDoc adheres to our team convention as discussed this morning. Also, next is not a type, it should be a function ie. `@param {Function} next - calls next middleware`

File: src/routes/auth.js [hunk 7]
```diff
@@ -1,9 +1,10 @@
 import express from 'express';
+
 import { signup, signin } from '../controllers/auth';
+import signupValidator from '../middlewares/AuthValidator';
```
Comment (iverenshaguy) [inline src/routes/auth.js]: We will need a validator middleware that will interface with the different validators for each route. Please look into this.

File: src/middlewares/AuthValidator.js [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    .checkBody('password')
+    .notEmpty()
+    .withMessage('Password field is required')
+    .trim()
+    .isLength({ min: 8, max: 15 })
+    .withMessage('Password should be between 8 to 15 characters')
+    .matches(/[aA-zZ]+[0-9]+/)
+    .withMessage('Password must be alphanumeric characters');
+
+  const errors = request.validationErrors();
```
Comment (iverenshaguy) [inline src/middlewares/AuthValidator.js]: I don't like this file. It needs to be refactored to have one validation middleware that validates all routes based on the different field validators.

File: tests/auth.test.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+  invalidLastName,
+  undefinedEmail,
+  undefinedUsername,
+  shortUsername,
+  invalidEmail,
+  undefinedPassword,
+  shortPassword,
+  invalidPassword,
+  validUser
+} from './mockData/mockUser';
```
Comment (iverenshaguy) [inline tests/auth.test.js]: Please change the `mockData` folder to the `fixtures` folder. Tests fixtures sound better than `mockData`.

File: tests/routes/auth.test.js [hunk 10]
```diff
@@ -21,7 +21,6 @@ describe('Auth Routes', () => {
         .request(app)
         .post('/api/v1/auth/signup')
         .send(user);
-
```
Comment (iverenshaguy) [inline tests/routes/auth.test.js]: Please leave the new line

**Review by iverenshaguy — CHANGES_REQUESTED** [review]
Please review my comments

File: tests/routes/auth.test.js (hunk 10 above)
Comment (Jonathan4github) [reply tests/routes/auth.test.js]: Thanks for the feedback working on it.

File: src/middlewares/validator.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ * Generic schema validator
+ *
+ * @param {Array} schema
+ * @param {number} status
+ *
+ * @returns {object} returns response message
+ */
+
+export default (schemas, status = 400) => {
+  const schemaTovalidate = schemas || [];
```
Comment (iverenshaguy) [inline src/middlewares/validator.js]: Really good job. Please update to 
```suggestion
  const schemaToValidate = schemas || [];
```

File: src/middlewares/validator.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
+import { validationResult } from 'express-validator/check';
+import { matchedData } from 'express-validator/filter';
+
+/**
+ * Generic schema validator
+ *
+ * @param {Array} schema
+ * @param {number} status
+ *
+ * @returns {object} returns response message
```
Comment (iverenshaguy) [inline src/middlewares/validator.js]:
```suggestion
 * @returns {Array} array of validation schema and middleware
```

File: src/middlewares/validator.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+  const schemaTovalidate = schemas || [];
+
+  const validationCheck = async (request, response, next) => {
+    const errors = validationResult(request);
+    request = { ...request, ...matchedData(request) };
+
+    if (!errors.isEmpty()) {
+      const mappedErrors = errors.mapped();
+
+      return response.status(status).json(mappedErrors);
```
Comment (iverenshaguy) [inline src/middlewares/validator.js]: What format is this error in? Please make sure it follows our convention and drop a snippet of the format here

File: src/routes/auth.js [hunk 14]
```diff
@@ -42,7 +46,9 @@ const router = express.Router();
  *       200:
  *         description: token is supplied
  */
-router.post('/signup', signup);
+
+
```
Comment (iverenshaguy) [inline src/routes/auth.js]: Any reason for the newlines after the documentation?

**Review by iverenshaguy — CHANGES_REQUESTED** [review]
You've done a fantastic job. Kindly review my comments so that we can merge

**Review by iverenshaguy — APPROVED** [review]

**Pull request merged**